// lobby-bot-3/bin/lib/language/UsageException.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsageException = exports.toUser = void 0;
const language_1 = require("../language");
function toUser(type, ...args) {
    return new UsageException(type, args);
}
exports.toUser = toUser;
class UsageException extends Error {
    constructor(type, args) {
        super(`UsageException: ${type}, ${language_1.configs[type].description}`);
        this.type = type;
        this.args = args;
        Object.setPrototypeOf(this, new.target.prototype);
    }
    async printMessage() {
        var _a, _b;
        const config = language_1.configs[this.type];
        return { embeds: [{
                    title: "Command Failed",
                    color: "RED",
                    description: config.description,
                    fields: (_b = (_a = config.fields) === null || _a === void 0 ? void 0 : _a.call(config)) !== null && _b !== void 0 ? _b : null,
                    footer: { text: "Use !help to learn more" },
                }] };
    }
}
exports.UsageException = UsageException;
