// lobby-bot-3/bin/lib/collections/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iter = exports.iterator = exports.shuffle = exports.Ref = exports.Reference = exports.GSet = exports.GetStrongSet = exports.FSet = exports.FastSet = exports.SSet = exports.StrongSet = void 0;
var fast_set_1 = require("./fast-set");
Object.defineProperty(exports, "StrongSet", { enumerable: true, get: function () { return fast_set_1.StrongSet; } });
Object.defineProperty(exports, "SSet", { enumerable: true, get: function () { return fast_set_1.StrongSet; } });
var fast_set_2 = require("./fast-set");
Object.defineProperty(exports, "FastSet", { enumerable: true, get: function () { return fast_set_2.FastSet; } });
Object.defineProperty(exports, "FSet", { enumerable: true, get: function () { return fast_set_2.FastSet; } });
var fast_set_3 = require("./fast-set");
Object.defineProperty(exports, "GetStrongSet", { enumerable: true, get: function () { return fast_set_3.GetStrongSet; } });
Object.defineProperty(exports, "GSet", { enumerable: true, get: function () { return fast_set_3.GetStrongSet; } });
var reference_1 = require("./reference");
Object.defineProperty(exports, "Reference", { enumerable: true, get: function () { return reference_1.Reference; } });
Object.defineProperty(exports, "Ref", { enumerable: true, get: function () { return reference_1.Reference; } });
var shuffle_1 = require("./shuffle");
Object.defineProperty(exports, "shuffle", { enumerable: true, get: function () { return shuffle_1.shuffle; } });
var iterator_1 = require("./iterator");
Object.defineProperty(exports, "iterator", { enumerable: true, get: function () { return iterator_1.iterator; } });
Object.defineProperty(exports, "iter", { enumerable: true, get: function () { return iterator_1.iterator; } });
